processing/src/main/java/io/druid/segment/incremental/OnheapIncrementalIndex.java

```diff
@@ -38,22 +38,31 @@
 import javax.annotation.Nullable;
 import java.io.IOException;
 import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
 import java.util.concurrent.ConcurrentHashMap;
 import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicLong;
 
 /**
  */
 public class OnheapIncrementalIndex extends IncrementalIndex<Aggregator>
 {
   private static final Logger log = new Logger(OnheapIncrementalIndex.class);
-
+  /**
+   *  overhead per Map$Entry object
+   *  KeyHash + next pointer + key pointer + value pointer + safe extra
+   */
+  private static final int ROUGH_OVERHEAD_PER_MAP_ENTRY = Long.BYTES * 5 + Integer.BYTES;
+  private static final long defaultMaxBytesInMemory = Runtime.getRuntime().maxMemory() / 3;
```

Comment:
I'd set the default for `OnheapIncrementalIndex` to -1 (unlimited).

The idea is that IncrementalIndex is the utility class and it is something of a blank slate. It's potentially used for more than just indexing: for example it's also used in groupBy v1. And in that case, this max of `maxMemory / 3` doesn't make sense, since there might be a lot more than just 1 index.

But at the _task_ level, in the tuning configs, it makes sense to have a default of `Runtime.getRuntime().maxMemory() / 3`. And in practice that's the value we'll go with at indexing time, since it overrides whatever is set as the IncrementalIndex default.

Another way of thinking about it: the default in IncrementalIndex "doesn't matter" since we expect it to get overridden. So it might as well be -1. And we don't want this magic code `Runtime.getRuntime().maxMemory() / 3` in more than 1 place, since it's magical, and magic should be rare in the world.

Reply:
I guess I was little unsure to set the default `maxMemory` here. Looking at how `maxRowCount` is handled, exception is thrown if its `<=0` in `IncrementalIndex.buildOnHeap` and `IncrementalIndex.buildOffHeap`. So I can either follow that and check `maxBytesInMemory <= 0` in the same places and throw exception (that might break some test cases, so will have to make the default >= 0 for `maxBytesInMemory` in those tests) or do `this.maxBytesInMemory = maxBytesInMemory == 0 ? -1 : maxBytesInMemory;` like you suggest here.